src/main/java/blusunrize/immersiveengineering/api/shader/ShaderLayer.java

```diff
@@ -0,0 +1,98 @@
+package blusunrize.immersiveengineering.api.shader;
+
+import net.minecraft.util.ResourceLocation;
+
+import javax.vecmath.Vector4f;
+
+public class ShaderLayer
+{
+	/**
+	 * A resource location pointing to a texture on the sheet
+	 */
+	private final ResourceLocation texture;
+	/**
+	 * An ARGB formatted colour
+	 */
+	private final Vector4f color;
+	/**
+	 * An optional double array (uMin, vMin, uMax, vMax; values of 0-1) to define which part of the original texture is overriden<br>
+	 * The model will then only render faces who's coords lie within that limited space.<br>
+	 * Useful for keeping additional decorations (see Sponsor Shader) in smaller textures
+	 */
+	private double[] textureBounds;
+	/**
+	 * An optional double array (uMin, vMin, uMax, vMax; values of 0-1) for the parts of the texture to be used.<br>
+	 * Useful when putting multiple shader textures into one file
+	 */
+	private double[] cutoutBounds;
+
+	public ShaderLayer(ResourceLocation texture, int color)
+	{
+		this.texture = texture;
+		this.color = new Vector4f((color >> 16&255)/255f, (color >> 8&255)/255f, (color&255)/255f, (color >> 24&255)/255f);
+		if(ShaderRegistry.defaultLayerBounds.containsKey(texture))
+			this.setTextureBounds(ShaderRegistry.defaultLayerBounds.get(texture));
+	}
+
+	public ShaderLayer setTextureBounds(double... bounds)
+	{
+		if(bounds==null)
+			return this;
+		assert (bounds.length==4);
```

Comment:
I prefer using `Preconditions.checkArgument` over `assert`s (at least for cheap things), since the former will always run and catch issues that might crash at any point early even in production (IIRC assert is disabled there by default).